### DISCUSSION TITLE
export as .stl by re-routing to three.js?

### DISCUSSION THREAD

**Discussion by rawwerks** [google/model-viewer#4562] [Q&A]
any thoughts on this approach as a workaround to export the current scene first to .glb and then to .stl?

(it's not working, currently getting `Uncaught (in promise) glTF export failed`)

better ideas for how to do this?

```
async function getSTLfromMV() {
		const modelViewer = document.getElementById("mv");
		const glTF = await modelViewer.exportScene();
		const blob = new Blob([glTF], {type: 'application/octet-stream'});
			let reader = new FileReader();
			reader.readAsArrayBuffer(blob);
			reader.onload = function () {
				let loader = new THREE.GLTFLoader();
				loader.parse(reader.result, '', function (gltf) {
				// Now you have a Three.js scene
				let scene = gltf.scene;

				// Use the STLExporter on the scene
				let exporter = new THREE.STLExporter();
				let stlString = exporter.parse(scene);

				// Now you have an STL string that you can download or do whatever with
				console.log(stlString);

				let stlBlob = new Blob([stlString], { type: 'application/sla' });
				let stlFile = new File([stlBlob], "model.stl");
				console.log(stlFile);

				});
			};
	}
```

**Comment by elalish** [comment]
Seems fine - set "break on uncaught exceptions" and try to figure out what's wrong. File an issue if you get a reproducible bug on `exportScene`.

**Reply by rawwerks** [reply]
thanks @elalish - can you clarify what you mean by `set "break on uncaught exceptions"` ?

here's the full error. i'm using modelviewer in a sveltekit app

```
Uncaught (in promise) glTF export failed [+page.svelte:691:1](http://127.0.0.1:5173/src/routes/+page.svelte)
    getSTLfromMV +page.svelte:691
    AsyncFunctionThrow self-hosted:856
    (Async: async)
    processGLB +page.svelte:302
    AsyncFunctionNext self-hosted:852
    (Async: async)
    onload +page.svelte:135
    (Async: EventHandlerNonNull)
    loadGLBFile +page.svelte:134
    uploaded +page.svelte:122
    dropHandler +page.svelte:480
```

**Reply by rawwerks** [reply]
if i try to call my ModelViewer instance directly...

```
import ModelViewer from '$lib/components/ModelViewer.svelte';
let mv: ModelViewer;
```

...i get this error instead: 
`Uncaught (in promise) TypeError: mv.exportScene is not a function`

(even though all of the other functions work this way... ie `mv.setModelUrl()`, `mv.exportImage()`, `mv.setModelSettings()` all work )

i am using: 
```
"@google/model-viewer": "^3.3.0",
"three": "^0.157.0"
```

**Reply by elalish** [reply]
Would help a lot if I could see your page live. I'm referring to using Chrome's dev tools debugger - are you familiar? What MV are you using? We have no APIs called `setModelUrl()` or `exportImage()`. https://modelviewer.dev/docs/index.html

**Comment by yosun** [comment]
hey any update on this?
